File: pythonCodeConversion/electricbill/controllers/customer_db_control.py
import mysql.connector


class CustomerDbControl:
    def __init__(self, connection):
        self.connection = connection

    def add_customer_to_database(self, customer):
        sql = "INSERT INTO customer (fName, nickname, gender, phoneNumber, address, email, subscriptionType, ampere, " \
              "status, balance) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"
        values = (
            customer.get_name(),
            customer.get_nickname(),
            customer.get_gender(),
            customer.get_phone_number(),
            customer.get_address(),
            customer.get_email(),
            customer.get_subscription_type(),
            customer.get_subscription_quantity(),
            customer.get_status(),
            customer.get_balance()
        )

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, values)
            self.connection.commit()
            print("Customer added successfully to the database!")
        except mysql.connector.Error as e:
            print("Error adding customer to the database:", e)
        finally:
            cursor.close()

    @staticmethod
    def display_table(data):
        num_columns = len(data)
        num_rows = (num_columns + 1) // 2

        for i in range(num_rows):
            row = data[i::num_rows]
            print("".join(f"{val:<20}" for val in row))

    def view_customers_from_database(self):
        sql = "SELECT idcustomer, fName, nickname, phoneNumber, subscriptionType, ampere, status, balance FROM customer"

        try:
            cursor = self.connection.cursor()
            cursor.execute(sql)
            result = cursor.fetchall()

            if not result:
                print("No customers found in the database.")
            else:
                print("Customers in the database:")
                headers = ["ID", "Name", "Phone Number", "Type", "Ampere", "Status", "Balance"]
                self.display_table(headers)

                for row in result:
                    customer_id, first_name, nickname, phone_number, subscription_type, ampere, status, balance = row
                    full_name = f"{first_name} {nickname}"
                    row_data = [
                        str(customer_id),
                        full_name,
                        phone_number,
                        subscription_type,
                        str(ampere),
                        status,
                        str(balance)
                    ]
                    self.display_table(row_data)
        except mysql.connector.Error as e:
            print("Error viewing customers from the database:", e)
        finally:
            cursor.close()

    def delete_customer_from_database(self, customer_id):
        sql = "DELETE FROM customer WHERE idcustomer = %s"
        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, (customer_id,))
            self.connection.commit()
            if cursor.rowcount > 0:
                print(f"Customer with ID {customer_id} deleted successfully.")
            else:
                print(f"No customer found with ID {customer_id}. No deletion performed.")
        except mysql.connector.Error as e:
            print("Error deleting customer from the database:", e)
        finally:
            cursor.close()

    def update_customer_info(self, customer_id, column_name, new_value):
        sql = f"UPDATE customer SET {column_name} = %s WHERE idcustomer = %s"
        try:
            cursor = self.connection.cursor()
            cursor.execute(sql, (new_value, customer_id))
            self.connection.commit()
            if cursor.rowcount > 0:
                print(f"Customer with ID {customer_id} updated successfully.")
            else:
                print(f"No customer found with ID {customer_id}. No update performed.")
        except mysql.connector.Error as e:
            print("Error updating customer info:", e)
        finally:
            cursor.close()

    @staticmethod
    def display_column_table():
        column_names = [
            "Name", "Nickname", "Gender", "Phone Number",
            "Address", "Email", "Subscription Type",
            "Ampere", "Status", "Balance"
        ]

        print("Column Index | Column Name")
        print("-------------|-------------")
        for i, column_name in enumerate(column_names, start=1):
            print(f"{i:<12} | {column_name}")

    def get_user_input_and_update(self):
        self.display_column_table()

        customer_id = int(input("\nEnter customer ID: "))
        column_index = int(input("Enter column index to update: "))
        new_value = input("Enter new value: ")

        column_name = self.get_column_name(column_index)
        self.update_customer_info(customer_id, column_name, new_value)

    @staticmethod
    def get_column_name(index):
        column_names = [
            "fName", "nickname", "gender", "phoneNumber",
            "address", "email", "subscriptionType",
            "ampere", "status", "balance"
        ]
        if 0 < index <= len(column_names):
            return column_names[index - 1]
        return ""

File: src/v5/CustomerDbControl.java
package v5;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;

public class CustomerDbControl {
    private final Connection connection;

    public CustomerDbControl(Connection connection) {
        this.connection = connection;
    }

    //Method to create new customer into the database
    public void addCustomerToDatabase(Customer customer) {
        String sql = "INSERT INTO customer (fName, nickname, gender, phoneNumber, address, email, subscriptionType, ampere, status, balance) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, customer.getName());
            preparedStatement.setString(2, customer.getNickname());
            preparedStatement.setString(3, customer.getGender());
            preparedStatement.setString(4, customer.getPhoneNumber());
            preparedStatement.setString(5, customer.getAddress());
            preparedStatement.setString(6, customer.getEmail());
            preparedStatement.setString(7, customer.getSubscriptionType());
            preparedStatement.setInt(   8, customer.getSubscriptionQuantity());
            preparedStatement.setString(9, customer.getStatus());
            preparedStatement.setDouble(10, customer.getBalance());

            preparedStatement.executeUpdate();
            System.out.println("Customer added successfully to the database!");
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Method to display table
    private static void displayTable(String[] data) {
        int numColumns = data.length;

        // Display table content
        int numRows = (int) Math.ceil((double) data.length / numColumns);
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numColumns; j++) {
                int index = i + j * numRows;
                if (index < data.length) {
                    System.out.printf("%-20s", data[index]);
                }
            }
            System.out.println();
        }
    }



    // Method to retrieve data as to table
    public void viewCustomersFromDatabase() {
        String sql = "SELECT idcustomer, fName, nickname, phoneNumber, subscriptionType, ampere, status, balance FROM customer";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {

            if (!resultSet.isBeforeFirst()) {
                System.out.println("No customers found in the database.");
            } else {
                System.out.println("Customers in the database:");

                // Table headers
                String[] headers = {"ID", "Name", "Phone Number", "Type", "Ampere", "Status", "Balance"};
                displayTable(headers);

                while (resultSet.next()) {
                    int customerId = resultSet.getInt("idcustomer");
                    String firstName = resultSet.getString("fName");
                    String lastName = resultSet.getString("nickName");
                    String phoneNumber = resultSet.getString("phoneNumber");
                    String subscriptionType = resultSet.getString("subscriptionType");
                    int ampere = resultSet.getInt("ampere");
                    String status = resultSet.getString("status");
                    double balance = resultSet.getDouble("balance");
                    double amount = subscriptionType.equals("24-Hours") ?ampere * 10000:ampere*5000;
                    balance= status.equals("Not Payed") ?amount:0;
                    // Formatted row data
                    String[] rowData = {
                            String.valueOf(customerId),
                            firstName + " " + lastName,
                            phoneNumber,
                            subscriptionType,
                            String.valueOf(ampere),
                            status,
                            String.valueOf(balance)
                    };

                    displayTable(rowData);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //method to delete customer in the database
    public void deleteCustomerFromDatabase(int customerId) {
        String sql = "DELETE FROM customer WHERE idcustomer = ?";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, customerId);

            int rowsAffected = preparedStatement.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Customer with ID " + customerId + " deleted successfully.");
            } else {
                System.out.println("No customer found with ID " + customerId + ". No deletion performed.");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //method to Update
    public void updateCustomerInfo(int customerId, String columnName, String newValue) {
        String sql = "UPDATE customer SET " + columnName + " = ? WHERE idcustomer = ?";

        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, newValue);
            preparedStatement.setInt(2, customerId);

            int rowsAffected = preparedStatement.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Customer with ID " + customerId + " updated successfully.");
            } else {
                System.out.println("No customer found with ID " + customerId + ". No update performed.");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    //DISPLAY
    public void displayColumnTable() {
        String[] columnNames = {
                "Name", "Nickname", "Gender", "Phone Number",
                "Address", "Email", "Subscription Type",
                "Ampere", "Status", "Balance"
        };

        System.out.println("Column Index | Column Name");
        System.out.println("-------------|-------------");
        for (int i = 0; i < columnNames.length; i++) {
            System.out.printf("%12d | %s\n", i + 1, columnNames[i]);
        }
    }

    // Method to get user input for ID, column index, and new value
    public void getUserInputAndUpdate() {
        Scanner scanner = new Scanner(System.in);
        CustomerDbControl custDbControl = new CustomerDbControl(connection);

        // Display the table of column names with index numbers
        displayColumnTable();

        // Get user input for ID, column index, and new value
        System.out.print("\nEnter customer ID: ");
        int customerId = scanner.nextInt();

        System.out.print("Enter column index to update: ");
        int columnIndex = scanner.nextInt();

        scanner.nextLine(); // Consume newline left from nextInt()

        System.out.print("Enter new value: ");
        String newValue = scanner.nextLine();

        // Call the update method with provided details
        custDbControl.updateCustomerInfo(customerId, getColumnName(columnIndex), newValue);
    }

    // Method to retrieve column name based on index
    private String getColumnName(int index) {
        String[] columnNames = {
                "fName", "nickname", "gender", "phoneNumber",
                "address", "email", "subscriptionType",
                "ampere", "status", "balance"
        };
        if (index > 0 && index <= columnNames.length) {
            return columnNames[index - 1];
        }
        return "";
    }

}
